File: observer/cpp/example_1/observer.cpp
#include <iostream>
#include <vector>

// like an interface in JAVA
class Observer
{
public:
  virtual ~Observer() {};
  virtual void update(float temp, float humidity, float pressure) = 0;
};

class Subject
{
public:
  virtual ~Subject() {};
  virtual void registerObserver(Observer* o) = 0;
};

class DisplayElement
{
public:
  virtual ~DisplayElement() {};
  virtual void display() const = 0;
};

class WeatherData: public Subject
{
private:
  std::vector<Observer*> observers;
  float temperature;
  float humidity;
  float pressure;

public:
  WeatherData() {
  }

  void registerObserver(Observer* o) override {
    this->observers.push_back(o);
  }

  void removeObserver(Observer* o) {
    // @todo
  }

  void notifyObservers() {
    for (auto observer : this->observers) {
      observer->update(temperature, humidity, pressure);
    }
  }

  void measurementsChanged() {
    this->notifyObservers();
  }

  void setMeasurements(float temperature_, float humidity_, float pressure_) {
    this->temperature = temperature_;
    this->humidity = humidity_;
    this->pressure = pressure_;
    this->measurementsChanged();
  }

  // other methods
};

class CurrentConditionsDisplay: public Observer, public DisplayElement
{
private:
  float temperature;
  float humidity;
  Subject* weatherData;

public:
  CurrentConditionsDisplay(Subject* weatherData_) {
    this->weatherData = weatherData_;
    weatherData->registerObserver(this);
  }

  void update(float temperature_, float humidity_, float pressure_) override {
    this->temperature = temperature_;
    this->humidity = humidity_;
    this->display();
  }

  void display() const override {
    std::cout << "Current condititions: " << this->temperature
              << "C degrees and " << this->humidity << "% humidity"
              << std::endl;
  }
};

int main() {
  WeatherData* weatherData = new WeatherData();
  CurrentConditionsDisplay* currentDisplay =
    new CurrentConditionsDisplay(weatherData);
  weatherData->setMeasurements(80, 65, 30.4f);
  delete currentDisplay;
  delete weatherData;
  return 0;
}

File: observer/python/example_1/observer.py
# -*- coding: utf-8 -*-

class Subject:

	def register_observer(observer):
		pass

	def remove_observer(observer):
		pass

	def notify_observers():
		pass


class Observer:

	def update(temp, humidity, pressure):
		pass


class DisplayElement:

	def display():
		pass


class WeatherData(Subject):

	_observers = []
	_temperature = 0.0
	_humidity = 0.0
	_pressure = 0.0

	def __init__(self):
		self._observers = []

	def register_observer(self, observer):
		self._observers.append(observer)

	def remove_observer(self, observer):
		try:
			i = self._observers.remove(observer)
		except ValueError:
			pass
	
	def notify_observers(self):
		for observer in self._observers:
			observer.update(self._temperature, self._humidity, self._pressure)

	def measurements_changed(self):
		self.notify_observers()

	def set_mesasurements(self, temperature, humidity, pressure):
		self._temperature = temperature
		self._humidity = humidity
		self._pressure = pressure
		self.measurements_changed()


class CurrentConditionsDisplay(Observer, DisplayElement):

	_temperature = 0.0
	_humidity = 0.0
	_weather_data = None

	def __init__(self, weather_data):
		self._weather_data = weather_data
		weather_data.register_observer(self)

	def update(self, temperature, humidity, pressure):
		self._temperature = temperature
		self._humidity = humidity
		self.display()

	def display(self):
		print("Current conditions: %s F degrees and %s humidity" % (self._temperature, self._humidity))

if __name__ == "__main__":
	weather_data = WeatherData()
	current_display = CurrentConditionsDisplay(weather_data)
	weather_data.set_mesasurements(80, 65, 30.4)
	weather_data.set_mesasurements(72, 25, 31.4)
	weather_data.set_mesasurements(31, 65, 32.4)
